File: apps/analytics/frontend/src/lib/predict.ts
/**
 * Predição XGBoost rodando 100% no browser via ONNX Runtime Web.
 *
 * Substitui a procedure `predictions.predict` do tRPC quando o site roda
 * estaticamente (sem backend Node). O modelo é carregado uma vez (cache
 * automático do navegador) e reaproveitado em chamadas subsequentes.
 *
 * Modelo:   /data/model/model.onnx        (~480 KB, XGBoost AUC 0.823)
 * Features: /data/model/features.json     (53 colunas + medianas para imputação)
 */
import * as ort from "onnxruntime-web";

// O onnxruntime-web precisa carregar arquivos WASM. Por padrão ele busca em
// `/`, o que falha quando o site fica num subpath (ex.: rftbrasil.com/analytics/).
// Apontar pra CDN evita ter que copiar os ~15 MB de WASM no nosso bundle.
ort.env.wasm.wasmPaths = "https://cdn.jsdelivr.net/npm/onnxruntime-web@1.25.1/dist/";

const MODEL_URL = "/data/model/model.onnx";
const FEATURES_URL = "/data/model/features.json";

type FighterInput = {
  name: string;
  wins?: number | null;
  losses?: number | null;
  draws?: number | null;
  age?: number | null;
  height_cm?: number | null;
  heightCm?: number | null;
  weight_kg?: number | null;
  weightKg?: number | null;
  reach_cm?: number | null;
  reachCm?: number | null;
  stance?: string | null;
  SLpM_total?: number | null;
  SLpM?: number | null;
  SApM_total?: number | null;
  SApM?: number | null;
  sig_str_acc_total?: number | null;
  td_acc_total?: number | null;
  str_def_total?: number | null;
  td_def_total?: number | null;
  sub_avg?: number | null;
  td_avg?: number | null;
  finishRate?: number | null;
  submissionWinRate?: number | null;
};

export type PredictionResult = {
  fighter1WinProbability: number;
  fighter2WinProbability: number;
  predictedWinner: string;
  confidence: number;
  fighter1DecimalOdds: number;
  fighter2DecimalOdds: number;
  fighter1AmericanOdds: number;
  fighter2AmericanOdds: number;
  keyFactors: string[];
  modelBreakdown: { model: string; f1Prob: number; f2Prob: number; weight: number }[];
  probKoTko: number;
  probSubmission: number;
  probDecision: number;
  fighter1KoProb: number;
  fighter1SubProb: number;
  fighter1DecProb: number;
  fighter2KoProb: number;
  fighter2SubProb: number;
  fighter2DecProb: number;
};

type FeatureCfg = { columns: string[]; medians: Record<string, number> };

let _session: ort.InferenceSession | null = null;
let _features: FeatureCfg | null = null;
let _loading: Promise<void> | null = null;

async function load(): Promise<void> {
  if (_session && _features) return;
  if (!_loading) {
    _loading = (async () => {
      const [session, featRes] = await Promise.all([
        ort.InferenceSession.create(MODEL_URL, {
          executionProviders: ["wasm"],
          graphOptimizationLevel: "all",
        }),
        fetch(FEATURES_URL).then((r) => r.json() as Promise<FeatureCfg>),
      ]);
      _session = session;
      _features = featRes;
    })();
  }
  await _loading;
}

const PER_FIGHTER = [
  "wins_total", "losses_total", "age", "height", "weight", "reach",
  "SLpM_total", "SApM_total", "sig_str_acc_total", "td_acc_total",
  "str_def_total", "td_def_total", "sub_avg", "td_avg",
];

const STANCES = ["Orthodox", "Southpaw", "Switch", "Open Stance"];

const slug = (s: string) => s.replace(/ /g, "_").toLowerCase();

const pick = (...vals: any[]): number | null => {
  for (const v of vals) if (v != null && !Number.isNaN(v)) return Number(v);
  return null;
};

function toFighterFeatures(f: FighterInput) {
  return {
    wins_total:        pick(f.wins),
    losses_total:      pick(f.losses),
    age:               pick(f.age),
    height:            pick(f.height_cm, f.heightCm),
    weight:            pick(f.weight_kg, f.weightKg),
    reach:             pick(f.reach_cm, f.reachCm),
    SLpM_total:        pick(f.SLpM_total, f.SLpM),
    SApM_total:        pick(f.SApM_total, f.SApM),
    sig_str_acc_total: pick(f.sig_str_acc_total),
    td_acc_total:      pick(f.td_acc_total),
    str_def_total:     pick(f.str_def_total),
    td_def_total:      pick(f.td_def_total),
    sub_avg:           pick(f.sub_avg),
    td_avg:            pick(f.td_avg),
    stance:            f.stance ?? "Orthodox",
  };
}

function buildRow(
  f1: FighterInput,
  f2: FighterInput,
  ctx: { is_title_bout?: number; total_rounds?: number; gender_male?: number },
  cfg: FeatureCfg,
): { values: Float32Array; named: Record<string, number> } {
  const r = toFighterFeatures(f1);
  const b = toFighterFeatures(f2);

  const feat: Record<string, number | null> = {};
  for (const k of PER_FIGHTER) {
    feat[`r_${k}`] = (r as any)[k];
    feat[`b_${k}`] = (b as any)[k];
  }
  for (const k of PER_FIGHTER) {
    const a = (r as any)[k];
    const c = (b as any)[k];
    feat[`${k}_diff`] = a != null && c != null ? Number(a) - Number(c) : null;
  }
  for (const stance of STANCES) {
    feat[`r_stance_${slug(stance)}`] = String(r.stance) === stance ? 1 : 0;
    feat[`b_stance_${slug(stance)}`] = String(b.stance) === stance ? 1 : 0;
  }
  feat["is_title_bout"] = Number(ctx.is_title_bout ?? 0);
  feat["total_rounds"]  = Number(ctx.total_rounds ?? 3);
  feat["gender_male"]   = Number(ctx.gender_male ?? 1);

  const values = new Float32Array(cfg.columns.length);
  const named: Record<string, number> = {};
  for (let i = 0; i < cfg.columns.length; i++) {
    const col = cfg.columns[i];
    let v = feat[col];
    if (v == null || Number.isNaN(v)) v = cfg.medians[col] ?? 0;
    values[i] = Number(v);
    named[col] = Number(v);
  }
  return { values, named };
}

const probaToDecimalOdds = (p: number) => Math.round((1 / Math.max(0.01, Math.min(0.99, p))) * 100) / 100;
const probaToAmericanOdds = (p: number) => {
  const x = Math.max(0.01, Math.min(0.99, p));
  return x >= 0.5 ? -Math.round((x / (1 - x)) * 100) : Math.round(((1 - x) / x) * 100);
};

function topFactors(named: Record<string, number>, k = 5): string[] {
  // O modelo ONNX não exporta feature_importances diretamente, então
  // priorizamos as features _diff (mais interpretáveis) com maior |valor|.
  const candidates = Object.entries(named)
    .filter(([col]) => col.endsWith("_diff"))
    .map(([col, v]) => ({ col, v, score: Math.abs(v) }))
    .sort((a, b) => b.score - a.score)
    .slice(0, k);
  return candidates.map(({ col, v }) => `${col} = ${v >= 0 ? "+" : ""}${v.toFixed(2)}`);
}

export async function predict(
  fighter1: FighterInput,
  fighter2: FighterInput,
  context: { is_title_bout?: number; total_rounds?: number; gender_male?: number } = {},
): Promise<PredictionResult> {
  await load();
  if (!_session || !_features) throw new Error("Model not loaded");

  const { values, named } = buildRow(fighter1, fighter2, context, _features);

  const tensor = new ort.Tensor("float32", values, [1, _features.columns.length]);
  const feeds = { [_session.inputNames[0]]: tensor };
  const out = await _session.run(feeds);

  // O output do XGBoost ONNX é tipicamente `probabilities` (mapa) ou um array com
  // [P(class=0), P(class=1)]. Detectamos qual está disponível.
  let pBlue = 0.5, pRed = 0.5;
  for (const name of _session.outputNames) {
    const t = out[name];
    if (!t) continue;
    const data = t.data as Float32Array | number[];
    if (data && data.length === 2) {
      pBlue = Number(data[0]);
      pRed  = Number(data[1]);
      break;
    }
  }

  // Heurística pra método (KO / Sub / Decisão) — modelo é binário winner-only
  const fr1 = Number(fighter1.finishRate ?? 0.5);
  const fr2 = Number(fighter2.finishRate ?? 0.5);
  const sub1 = Number(fighter1.submissionWinRate ?? 0);
  const sub2 = Number(fighter2.submissionWinRate ?? 0);
  const pFinishOverall = Math.max(0.05, Math.min(0.95, fr1 * pRed + fr2 * pBlue));
  const f1Sub = sub1 * pRed;
  const f1Ko  = Math.max(pFinishOverall * pRed - f1Sub, 0);
  const f2Sub = sub2 * pBlue;
  const f2Ko  = Math.max(pFinishOverall * pBlue - f2Sub, 0);
  const f1Dec = Math.max(pRed - f1Ko - f1Sub, 0);
  const f2Dec = Math.max(pBlue - f2Ko - f2Sub, 0);

  const winner = pRed >= pBlue ? fighter1.name : fighter2.name;
  const confidence = Math.max(pRed, pBlue);

  return {
    fighter1WinProbability: pRed,
    fighter2WinProbability: pBlue,
    predictedWinner: winner,
    confidence,
    fighter1DecimalOdds:  probaToDecimalOdds(pRed),
    fighter2DecimalOdds:  probaToDecimalOdds(pBlue),
    fighter1AmericanOdds: probaToAmericanOdds(pRed),
    fighter2AmericanOdds: probaToAmericanOdds(pBlue),
    keyFactors: topFactors(named, 5),
    modelBreakdown: [
      { model: "XGBoost (ONNX, browser)", f1Prob: pRed, f2Prob: pBlue, weight: 1.0 },
    ],
    probKoTko:      f1Ko + f2Ko,
    probSubmission: f1Sub + f2Sub,
    probDecision:   f1Dec + f2Dec,
    fighter1KoProb:  f1Ko,
    fighter1SubProb: f1Sub,
    fighter1DecProb: f1Dec,
    fighter2KoProb:  f2Ko,
    fighter2SubProb: f2Sub,
    fighter2DecProb: f2Dec,
  };
}

/**
 * Hook compatível com `usePredictMutation()`.
 *
 * Uso: trocar `usePredictMutation()` por `usePredictMutation()`.
 */
import { useMutation, type UseMutationOptions } from "@tanstack/react-query";

type PredictInput = {
  fighter1Id?: number;
  fighter2Id?: number;
  fighter1Name?: string;
  fighter2Name?: string;
  fighter1Data?: FighterInput;
  fighter2Data?: FighterInput;
  context?: { is_title_bout?: number; total_rounds?: number; gender_male?: number };
};

export function usePredictMutation(
  options?: Omit<UseMutationOptions<PredictionResult, Error, PredictInput>, "mutationFn">,
) {
  return useMutation<PredictionResult, Error, PredictInput>({
    ...options,
    mutationFn: async (input: PredictInput) => {
      // Resolve dados do atleta a partir do JSON estático se só vier o id
      const resolveFighter = async (id?: number, name?: string, override?: FighterInput): Promise<FighterInput> => {
        if (override) return override;
        if (id) {
          try {
            const r = await fetch(`/data/fighters/${id}.json`);
            if (r.ok) {
              const f = await r.json();
              return { ...f, name: f.name ?? name ?? "Fighter" };
            }
          } catch {}
        }
        return { name: name ?? "Fighter" };
      };

      const f1 = await resolveFighter(input.fighter1Id, input.fighter1Name, input.fighter1Data);
      const f2 = await resolveFighter(input.fighter2Id, input.fighter2Name, input.fighter2Data);
      return predict(f1, f2, input.context ?? {});
    },
  });
}

File: apps/analytics/backend/python/modulo_3_ml/predict.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Inference CLI — XGBoost predictor.

Recebe via stdin um JSON com os 2 atletas. Retorna no stdout um JSON com a
probabilidade de vitória, odds decimais/americanas, top features e breakdown.

Input esperado (stdin):
{
  "fighter1": {
     "name": "Petr Yan",
     "wins": 17, "losses": 6, "draws": 0,
     "age": 32, "height_cm": 170, "weight_kg": 61.2, "reach_cm": 170,
     "stance": "Orthodox",
     "SLpM_total": 4.6, "SApM_total": 3.2,
     "sig_str_acc_total": 0.5, "td_acc_total": 0.4,
     "str_def_total": 0.6, "td_def_total": 0.7,
     "sub_avg": 0.3, "td_avg": 1.5
  },
  "fighter2": { ... mesmo formato ... },
  "context": {
     "is_title_bout": 0, "total_rounds": 3, "gender_male": 1
  }
}

Output (stdout): JSON com a predição.

Uso:
  echo '<json>' | python3 predict.py
"""

import sys
import json
import logging
import math
from pathlib import Path
from typing import Any

import numpy as np
import pandas as pd
import joblib

logging.basicConfig(level=logging.WARNING, format="[%(levelname)s] %(message)s")

MODEL_DIR = Path(__file__).resolve().parent
MODEL_PATH = MODEL_DIR / "model_predictor.pkl"
COLUMNS_PATH = MODEL_DIR / "feature_columns.json"

# Carrega na importação (cacheia em invocações repetidas se for daemonizado)
_model = joblib.load(MODEL_PATH)
_cfg = json.loads(COLUMNS_PATH.read_text())
COLUMNS: list[str] = _cfg["columns"]
MEDIANS: dict[str, float] = _cfg["medians"]

# Mapeamento entre as features do modelo e o JSON de entrada.
PER_FIGHTER = [
    "wins_total", "losses_total", "age", "height", "weight", "reach",
    "SLpM_total", "SApM_total", "sig_str_acc_total", "td_acc_total",
    "str_def_total", "td_def_total", "sub_avg", "td_avg",
]
DIFF_FEATURES = [
    "wins_total_diff", "losses_total_diff", "age_diff", "height_diff",
    "weight_diff", "reach_diff", "SLpM_total_diff", "SApM_total_diff",
    "sig_str_acc_total_diff", "td_acc_total_diff", "str_def_total_diff",
    "td_def_total_diff", "sub_avg_diff", "td_avg_diff",
]
STANCE_VALUES = ["Orthodox", "Southpaw", "Switch", "Open Stance"]


def _slug(s: str) -> str:
    return s.replace(" ", "_").lower()


def _g(d: dict, *keys, default=None):
    """Pega o primeiro valor não-nulo de várias chaves alternativas."""
    for k in keys:
        v = d.get(k)
        if v is not None and not (isinstance(v, float) and math.isnan(v)):
            return v
    return default


def fighter_to_features(f: dict) -> dict:
    """Mapeia o JSON de input (um fighter) → features per-fighter (sem prefixo)."""
    return {
        "wins_total":          _g(f, "wins_total", "wins"),
        "losses_total":        _g(f, "losses_total", "losses"),
        "age":                 _g(f, "age"),
        "height":              _g(f, "height", "height_cm", "heightCm"),
        "weight":              _g(f, "weight", "weight_kg", "weightKg"),
        "reach":               _g(f, "reach", "reach_cm", "reachCm"),
        "SLpM_total":          _g(f, "SLpM_total", "SLpM"),
        "SApM_total":          _g(f, "SApM_total", "SApM"),
        "sig_str_acc_total":   _g(f, "sig_str_acc_total", "Str_Acc"),
        "td_acc_total":        _g(f, "td_acc_total", "TD_Acc"),
        "str_def_total":       _g(f, "str_def_total", "Str_Def"),
        "td_def_total":        _g(f, "td_def_total", "TD_Def"),
        "sub_avg":             _g(f, "sub_avg", "Sub_Avg"),
        "td_avg":              _g(f, "td_avg", "TD_Avg"),
        "stance":              _g(f, "stance", default="Orthodox"),
    }


def _safe_diff(a, b):
    if a is None or b is None:
        return None
    return float(a) - float(b)


def build_row(payload: dict) -> tuple[pd.DataFrame, dict]:
    """Constrói uma linha de features alinhada com COLUMNS (com imputação por mediana)."""
    f1 = fighter_to_features(payload["fighter1"])
    f2 = fighter_to_features(payload["fighter2"])
    ctx = payload.get("context", {}) or {}

    feat: dict[str, Any] = {}
    for k in PER_FIGHTER:
        feat[f"r_{k}"] = f1.get(k)
        feat[f"b_{k}"] = f2.get(k)

    for d in DIFF_FEATURES:
        base = d.replace("_diff", "")
        feat[d] = _safe_diff(f1.get(base), f2.get(base))

    for stance in STANCE_VALUES:
        feat[f"r_stance_{_slug(stance)}"] = 1 if str(f1.get("stance")) == stance else 0
        feat[f"b_stance_{_slug(stance)}"] = 1 if str(f2.get("stance")) == stance else 0

    feat["is_title_bout"] = int(ctx.get("is_title_bout", 0) or 0)
    feat["total_rounds"]  = float(ctx.get("total_rounds", 3) or 3)
    feat["gender_male"]   = int(ctx.get("gender_male", 1) or 0)

    # Imputação com mediana onde estiver vazio
    row = {}
    for col in COLUMNS:
        v = feat.get(col)
        if v is None or (isinstance(v, float) and math.isnan(v)):
            v = MEDIANS.get(col)
        row[col] = v

    df = pd.DataFrame([row], columns=COLUMNS)
    return df, feat


def proba_to_decimal_odds(p: float) -> float:
    p = max(min(p, 0.99), 0.01)
    return round(1.0 / p, 2)


def proba_to_american_odds(p: float) -> int:
    p = max(min(p, 0.99), 0.01)
    if p >= 0.5:
        return -int(round((p / (1 - p)) * 100))
    return int(round(((1 - p) / p) * 100))


def top_features(model, df: pd.DataFrame, k: int = 5) -> list[str]:
    """Retorna top features por |valor × importance|, com sinal explicativo."""
    importances = getattr(model, "feature_importances_", None)
    if importances is None:
        return []
    row_vals = df.iloc[0].values
    contribs = []
    for i, col in enumerate(df.columns):
        v = row_vals[i]
        if v is None or (isinstance(v, float) and math.isnan(v)):
            continue
        score = abs(float(v)) * float(importances[i])
        contribs.append((score, col, float(v)))
    contribs.sort(reverse=True)
    out = []
    for _, col, v in contribs[:k]:
        if "_diff" in col:
            out.append(f"{col} = {v:+.2f}")
        else:
            out.append(f"{col} = {v}")
    return out


def main():
    payload = json.loads(sys.stdin.read())
    df, _ = build_row(payload)

    proba = _model.predict_proba(df)[0]
    p_red = float(proba[1])
    p_blue = float(proba[0])

    f1_name = payload.get("fighter1", {}).get("name", "Fighter 1")
    f2_name = payload.get("fighter2", {}).get("name", "Fighter 2")

    winner = f1_name if p_red >= p_blue else f2_name
    confidence = max(p_red, p_blue)

    # Heurística simples para método (KO / Sub / Decisão) — modelo é binário winner-only
    fr1 = float(payload["fighter1"].get("finishRate", 0.5) or 0.5)
    fr2 = float(payload["fighter2"].get("finishRate", 0.5) or 0.5)
    sub1 = float(payload["fighter1"].get("submissionWinRate", 0.0) or 0.0)
    sub2 = float(payload["fighter2"].get("submissionWinRate", 0.0) or 0.0)
    p_finish_overall = max(0.05, min(0.95, (fr1 * p_red + fr2 * p_blue)))
    p_dec_overall   = 1 - p_finish_overall
    f1_sub = sub1 * p_red + (1 - p_red) * 0  # rough
    f1_ko = max(p_finish_overall * p_red - f1_sub, 0)
    f2_sub = sub2 * p_blue
    f2_ko = max(p_finish_overall * p_blue - f2_sub, 0)
    f1_dec = max(p_red - f1_ko - f1_sub, 0)
    f2_dec = max(p_blue - f2_ko - f2_sub, 0)

    out = {
        "fighter1WinProbability": p_red,
        "fighter2WinProbability": p_blue,
        "predictedWinner": winner,
        "confidence": float(confidence),
        "fighter1DecimalOdds": proba_to_decimal_odds(p_red),
        "fighter2DecimalOdds": proba_to_decimal_odds(p_blue),
        "fighter1AmericanOdds": proba_to_american_odds(p_red),
        "fighter2AmericanOdds": proba_to_american_odds(p_blue),
        "keyFactors": top_features(_model, df, k=5),
        "modelBreakdown": [
            {"model": "XGBoost (large_dataset 94-24)", "f1Prob": p_red, "f2Prob": p_blue, "weight": 1.0},
        ],
        "probKoTko":     float(f1_ko + f2_ko),
        "probSubmission":float(f1_sub + f2_sub),
        "probDecision":  float(f1_dec + f2_dec),
        "fighter1KoProb":  float(f1_ko),
        "fighter1SubProb": float(f1_sub),
        "fighter1DecProb": float(f1_dec),
        "fighter2KoProb":  float(f2_ko),
        "fighter2SubProb": float(f2_sub),
        "fighter2DecProb": float(f2_dec),
    }

    sys.stdout.write(json.dumps(out))
    sys.stdout.flush()


if __name__ == "__main__":
    main()
